### PR TITLE
feat(codepipeline): add experimental support for the BitBucket source action

### LINKED ISSUE #6710: AWS CDK C#  Is it possible to add BitBucket as a source action in the CodePipeline in the source stage?
### The Question

Hello, I'm trying to link BitBucket as a source in my CodePipeline using Aws CDK C #, but I'm not getting it and I didn't find any documentation related to this item.

Has anyone managed to implement this?

Below example of the code I am using.

### Environment

  - **CDK CLI Version:** 1.27.0 (build a98c0b3)
  - **Module Version:** 
  - **OS:**  Windows 10
  - **Language:** csharp

### Other information

```csharp
public class CodePipelineStack : Stack
    {
        internal CodePipelineStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            var name = "name";
            var environment = Util.GetEnvironment(this.Account);
            var branch = "develop";
            var repo = "repo";

            var role = new Role(this, $"{name}-build-role", new RoleProps
            {
                RoleName = $"{name}-build-role",
                AssumedBy = new ServicePrincipal("codebuild.amazonaws.com"),
                Description = $"This is a custom role ({name}-build-role)",
                InlinePolicies = new Dictionary<string, PolicyDocument>
                {
                    [$"{name}-code-build-role"] = Util.ecsServicePolicies.Where(p => p.Key == "code-build-role").Select(s => s.Value).FirstOrDefault()
                }
            });
            

            var build = new PipelineProject(this, $"{name}-build", new PipelineProjectProps
            {
                BuildSpec = BuildSpec.FromSourceFilename($"buildspec-{environment}"),
                Environment = new BuildEnvironment
                {
                    BuildImage = LinuxBuildImage.STANDARD_1_0,
                    Privileged = true,
                    ComputeType = ComputeType.SMALL
                },
                Description = "Build Description",
                ProjectName = $"buildspec-{environment}",
                Timeout = Duration.Minutes(15),
                EnvironmentVariables = new Dictionary<string, IBuildEnvironmentVariable>
                {
                    ["ENVIRONMENT"] = new BuildEnvironmentVariable
                    {
                        Type = BuildEnvironmentVariableType.PLAINTEXT,
                        Value = environment
                    },
                    ["APP_NAME"] = new BuildEnvironmentVariable
                    {
                        Type = BuildEnvironmentVariableType.PLAINTEXT,
                        Value = name
                    }
                },
                Badge = false,
                Role = role,
                Cache = Cache.Bucket(new Bucket(this, $"{name}-build-cache",
                                        new BucketProps
                                        {
                                            Versioned = false,
                                            BucketName = $"{name}-build-cache",
                                            Encryption = BucketEncryption.KMS_MANAGED,
                                            RemovalPolicy = RemovalPolicy.DESTROY,
                                            BlockPublicAccess = new BlockPublicAccess(
                                                new BlockPublicAccessOptions
                                                {
                                                    BlockPublicAcls = true,
                                                    BlockPublicPolicy = true,
                                                    IgnorePublicAcls = true,
                                                    RestrictPublicBuckets = true
                                                }
                                            ),
                                            AccessControl = BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
                                            PublicReadAccess = false,
                                        }),
                                    new BucketCacheOptions
                                    {
                                        Prefix = environment
                                    })
            });

            var buildOutput = new Artifact_($"{name}-build-output");
            var sourceOutput = new Artifact_();

            var source = Amazon.CDK.AWS.CodeBuild.Source.BitBucket(new BitBucketSourceProps
            {
                BranchOrRef = "",
                CloneDepth = 0,
                Identifier = "",
                Owner = "",
                Repo = repo,
                ReportBuildStatus = true,
                Webhook = true,
                WebhookFilters = new[] {
                    FilterGroup.InEventOf(EventAction.PUSH).AndBranchIs(branch),
                    FilterGroup.InEventOf(EventAction.PULL_REQUEST_MERGED).AndBranchIs(branch),
                    FilterGroup.InEventOf(EventAction.PULL_REQUEST_UPDATED).AndBranchIs(branch),
                },
            });

            var pipeline = new Pipeline(this, $"{name}-pipeline", new PipelineProps
            {
                PipelineName = $"{name}-pipeline",
                RestartExecutionOnUpdate = true,
                Stages = new[]
                {
                    new StageProps
                    {
                        StageName = "Source",
                        Actions = new []
                        {
                            source // error in this line 
                        }
                    },
                    new StageProps
                    {
                        StageName = "Build",
                        Actions = new CodeBuildAction[]
                        {
                            new CodeBuildAction(new CodeBuildActionProps
                            {
                                ActionName = $"{name}-build",
                                Project = build,
                                Input = sourceOutput,
                                Outputs = new [] { buildOutput }
                            })
                        }
                    },
                }
            });
   }
    }
```

### PR DESCRIPTION
**Pull request by skinny85**
Fixes #6710

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline-actions/README.md [hunk 1]
```diff
@@ -106,6 +106,37 @@ new codepipeline_actions.CodeBuildAction({
...
+This is a one-time operation for a given AWS account in a given region.
+The simplest way to do that is to either start creating a new CodePipeline,
+or edit na existing one, while being logged in to BitBucket.
+Choose BitBucket as the source,
+and grant CodePipeline permissions to your BitBucket account.
+Copy & paste the Connection ARN that you get in the console,
+or use the [`codestar-connections list-connections` AWS CLI operation](https://docs.aws.amazon.com/cli/latest/reference/codestar-connections/list-connections.html)
+to find it.
+After that, you can safely abort creating or editing the pipeline -
+the connection has already been created.
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/README.md]: This is a pretty lousy user experience... Is there a plan to improve this?

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  readonly output: codepipeline.Artifact;
+
+  /**
+   * The ARN of the CodeStar Connection created in the AWS console
+   * that has permissions to access this BitBucket repository.
+   *
+   * @example 'arn:aws:codestar-connections:us-east-1:123456789012:connection/12345678-abcd-12ab-34cdef5678gh'
+   */
+  readonly connection: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]:
```suggestion
  readonly connectionArn: string;
```

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+export interface BitBucketSourceActionProps extends codepipeline.CommonAwsActionProps {
+  /**
+   * The output artifact that this action produces.
+   * Can be used as input for further pipeline actions.
+   */
+  readonly output: codepipeline.Artifact;
+
+  /**
+   * The ARN of the CodeStar Connection created in the AWS console
+   * that has permissions to access this BitBucket repository.
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: Reference to AWS Docs on how to create the connection?

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+   *
+   * @default 'master'
+   */
+  readonly branch?: string;
+
+  /**
+   * Whether the output should be the contents of the repository
+   * (which is the default),
+   * or a link that allows CodeBuild to clone the repository before building.
+   *
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: Is there any AWS docs reference that explains this feature. I couldn't understand what this does or what is the use case that this addresses from the description here...

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+   * @example 'aws'
+   */
+  readonly owner: string;
+
+  /**
+   * The name of the repository.
+   *
+   * @example 'aws-cdk'
+   */
+  readonly repo: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: I don't know that we need to split `owner` and `repo` (is it even called "owner" in bit bucket's terminology). As a user I would rather just specify `repo: "aws/aws-cdk"). We can add validation that this is the format.

File: packages/@aws-cdk/aws-codepipeline-actions/README.md (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/README.md]: I plan to reach out to the team owning this to talk about a way to make it automated.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: Done.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: Done.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: I don't fully understand it either, honestly, and I couldn't get it to work. Will add a link.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: The thing is, in all places you can provide this ([CodePipeline GitHub source](https://github.com/aws/aws-cdk/blob/a2a2f130ffec176a01991376679eeccf1027707e/packages/%40aws-cdk/aws-codepipeline-actions/lib/github/source-action.ts#L44-L52), [CodeBuild GitHub source](https://github.com/aws/aws-cdk/blob/a2a2f130ffec176a01991376679eeccf1027707e/packages/%40aws-cdk/aws-codebuild/lib/source.ts#L576-L588) and [CodeBuild BitBucket source](https://github.com/aws/aws-cdk/blob/a2a2f130ffec176a01991376679eeccf1027707e/packages/%40aws-cdk/aws-codebuild/lib/source.ts#L665-L677)), we take it in as an `owner` / `repo` pair - so I would rather keep it consistent across our libraries).

**Comment by skinny85** [comment]
@eladb thanks for the review! Updated the PR with changes from your comments.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts (hunk 5 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/bitbucket/source-action.ts]: Ok

**Pull request merged**